File: Data Structures & Algorithms/subsets/submission-0.java
class Solution {
    
    public List<List<Integer>> subsets(int[] nums) {

        List<List<Integer>> res = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();
        dfs(nums, 0, subset, res);
        return res;
    }

    private void dfs(int[] nums, int i, List<Integer> subset, List<List<Integer>> res) {
        
        if ( i >= nums.length) {
            res.add(new ArrayList<>(subset));
            return;
        }

        subset.add(nums[i]);
        dfs(nums, i + 1, subset, res);
        subset.remove(subset.size() - 1);
        dfs(nums, i + 1, subset, res);


    }

}


// class Solution {
//     public List<List<Integer>> subsets(int[] nums) {

//         List<List<Integer>> res = new ArrayList<>();
//         res.add(new ArrayList<>());

//         for ( int num : nums) {
//             int size = res.size();
//             for ( int i = 0; i < size; i ++) {
//                 List<Integer> subset = new ArrayList<>(res.get(i));
//                 subset.add(num);
//                 res.add(subset);
//             }  
//         }
//         return res;
//     }
// }

File: Data Structures & Algorithms/non-cyclical-number/submission-0.py
class Solution:
    def isHappy(self, n: int) -> bool:

        visit = set()

        while n not in visit:
            visit.add(n)
            n = self.sumOfSquares(n)

            if n == 1:
                return True

        return False

    def sumOfSquares(self, n: int) -> int:

            res = 0
            ns = str(n)

            for c in ns:
                res += int(c)*int(c)
            return res




        